stablehlo/tests/interpret_probe.mlir

```diff
@@ -0,0 +1,84 @@
+  check.expect_serialized_eq %3 { probe_id = "probe_i1" } : tensor<3x1xi1>
+  func.return
+}
+
+// -----
+
+func.func @probe_si64() {
+  %0 = stablehlo.constant dense<[[-127], [126], [0]]> : tensor<3x1xi64>
+  %1 = stablehlo.constant dense<[[-1], [1], [1]]> : tensor<3x1xi64>
+  %2 = stablehlo.add %0, %1 : tensor<3x1xi64>
+  %3 = interpreter.probe %2 { probe_id = "probe_si64" } : (tensor<3x1xi64>) -> tensor<3x1xi64>
+  check.expect_serialized_eq %3 { probe_id = "probe_si64" } : tensor<3x1xi64>
+  func.return
+}
+
+// -----
+
+func.func @probe_ui64() {
+  %0 = stablehlo.constant dense<[[32766, 0], [0, 0]]> : tensor<2x2xui64>
+  %1 = stablehlo.constant dense<[[1, 1], [2, 2]]> : tensor<2x2xui64>
+  %2 = stablehlo.add %0, %1 : tensor<2x2xui64>
+  %3 = interpreter.probe %2 { probe_id = "probe_ui64" } : (tensor<2x2xui64>) -> tensor<2x2xui64>
+  check.expect_serialized_eq %3 { probe_id = "probe_ui64" } : tensor<2x2xui64>
+  func.return
+}
+
+// -----
+
+func.func @probe_f64() {
+  %0 = stablehlo.constant dense<[3.402e+38, 1.175e-38, 0.0e+00]> : tensor<3xf64>
+  %1 = stablehlo.constant dense<[1.0e+00, -1.0e+00, 1.0e+00]> : tensor<3xf64>
+  %2 = stablehlo.add %0, %1 : tensor<3xf64>
+  %3 = interpreter.probe %2 { probe_id = "probe_f64" } : (tensor<3xf64>) -> tensor<3xf64>
+  check.expect_serialized_eq %3 { probe_id = "probe_f64" } : tensor<3xf64>
+  func.return
+}
+
+// -----
+
+func.func @probe_c32() {
+  %0 = stablehlo.constant dense<[(1.0, 2.0), (3.0, 4.0), (5.0, 6.0)]> : tensor<3xcomplex<f32>>
+  %1 = stablehlo.constant dense<[(1.0, 1.0), (1.0, 1.0), (1.0, 1.0)]> : tensor<3xcomplex<f32>>
+  %2 = stablehlo.add %0, %1 : tensor<3xcomplex<f32>>
+  %3 = interpreter.probe %2 { probe_id = "probe_c32" } : (tensor<3xcomplex<f32>>) -> tensor<3xcomplex<f32>>
+  check.expect_serialized_eq %3 { probe_id = "probe_c32" } : tensor<3xcomplex<f32>>
+  func.return
+}
+
+// -----
+
+func.func @probe_iterations() {
+  // int i = 0;
+  // int sum = 0;
+  // while (i < 10) {
+  //   sum += 1;
+  //   i += 1;
+  // }
+  %init_i = stablehlo.constant dense<0> : tensor<i64>
+  %init_sum = stablehlo.constant dense<0> : tensor<i64>
+  %one = stablehlo.constant dense<1> : tensor<i64>
+  %two = stablehlo.constant dense<2> : tensor<i64>
+  %results0, %results1 = stablehlo.while(%arg0 = %init_i, %arg1 = %init_sum) : tensor<i64>, tensor<i64>
+  cond {
+    %cond = stablehlo.compare LT, %arg0, %two : (tensor<i64>, tensor<i64>) -> tensor<i1>
+    stablehlo.return %cond : tensor<i1>
+  } do {
+    %new_sum = stablehlo.add %arg1, %one : tensor<i64>
+    %new_sum_instrumented = interpreter.probe %new_sum { probe_id = "probe_iterations" } : (tensor<i64>) -> tensor<i64>
+    %new_i = stablehlo.add %arg0, %one : tensor<i64>
+    stablehlo.return %new_i, %new_sum_instrumented : tensor<i64>, tensor<i64>
+  }
+
+  check.expect_eq_const %results0, dense<2> : tensor<i64>
+  check.expect_serialized_eq %one { probe_id = "probe_iterations", iteration = 0 : ui32 } : tensor<i64>
```

Comment:
Similarly (but also optional):
`check.expect_serialized_eq %one { probe_id = "probe_iterations", iteration = 0 : ui32 } : tensor<i64>`
into:
`check.expect_serialized_eq %one, "probe_iterations", i = 0 : tensor<i64>`

Something like:
```
let assemblyFormat = [{
  $operand `,` $expected `,` $probe_id
  (`,` `i` `=` $iteration^)?
  attr-dict `:` type($expected)
}];
```

Reply:
Two (undocumented) design principles for pretty printing:
1. Prefer N=V format as a default, omitting `Name=` is an exception for cases where it is overly redundant, like equality comparison type (`EQ`) for the sake of looking like `arith` dialect printing. Would suggest `id = $probe_id`
2. Abbreviations are fine, but prefer something that indicates the field name (i=index? i=iteration?) so would prefer `iter`, similar to how we use `dims` instead of `d`.

```
let assemblyFormat = [{
  $operand `,` $expected `,` `id` `=` $probe_id
  (`,` `iter` `=` $iteration^)?
  attr-dict `:` type($expected)
}];
```

Reply:
I  slightly tweaked the proposed `td` changes to have the attribute on `check.expect_serialized_eq` be `probe_id` instead to imply it's to be used with probe instrumentation.